File: 1392-longest-happy-prefix/1392-longest-happy-prefix.rb
# @param {String} s
# @return {String}
def longest_prefix(s)
  n = s.length
  lps = Array.new(n, 0) 

  length = 0  
  i = 1
  while i < n
    if s[i] == s[length]
      length += 1
      lps[i] = length
      i += 1
    else
      if length != 0
        length = lps[length - 1]
      else
        lps[i] = 0
        i += 1
      end
    end
  end

  s[0...lps[-1]]
end

File: 1392-longest-happy-prefix/1392-longest-happy-prefix.cpp
class Solution {
public:
    string longestPrefix(string s) {
        int n = s.size();
        vector<int> lps(n, 0);

        int length = 0;  
        int i = 1;
        while (i < n) {
            if (s[i] == s[length]) {
                length++;
                lps[i] = length;
                i++;
            } else {
                if (length != 0) {
                    length = lps[length - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }

        return s.substr(0, lps[n-1]);
    }
};
